hsds/attr_dn.py

```diff
@@ -203,69 +265,90 @@ async def PUT_Attribute(request):
-        log.error("PUT attribute with no type in body")
-        raise HTTPInternalServerError()
-
-    datatype = body["type"]
-
-    if "shape" not in body:
-        log.error("PUT attribute with no shape in body")
-        raise HTTPInternalServerError()
-    shape = body["shape"]
+    if "attributes" in body:
+        items = body["attributes"]
+    else:
+        # make it look like a dictionary anyway to make
+        # the processing more consistent
+        items = {}
+        if "name" not in body:
+            log.error("PUT attribute with no name in body")
+            raise HTTPInternalServerError()
+        attr_name = body["name"]
+        attribute = {}
+        if "type" in body:
+            attribute["type"] = body["type"]
+        if "shape" in body:
+            attribute["shape"] = body["shape"]
+        if "value" in body:
+            attribute["value"] = body["value"]
+        items[attr_name] = attribute
+    
+    # validate input
+    for attr_name in items:
+        validateAttributeName(attr_name)
+        attr_json = items[attr_name]
+        if "type" not in attr_json:
+            log.error("PUT attribute with no type in body")
+            raise HTTPInternalServerError()
+        if "shape" not in attr_json:
+            log.error("PUT attribute with no shape in body")
+            raise HTTPInternalServerError()
 
-    if "value" in body:
-        value = body["value"]
+    log.info(f"PUT {len(items)} attributes to obj_id: {obj_id} bucket: {bucket}")
 
     obj_json = await get_metadata_obj(app, obj_id, bucket=bucket)
-    log.debug(f"PUT attribute obj_id: {obj_id} bucket: {bucket} got json")
-
     if "attributes" not in obj_json:
         log.error(f"unexpected obj data for id: {obj_id}")
         raise HTTPInternalServerError()
 
     attributes = obj_json["attributes"]
-    if attr_name in attributes and not replace:
-        # Attribute already exists, return a 409
-        msg = f"Attempt to overwrite attribute: {attr_name} "
-        msg += f"in obj_id: {obj_id}"
-        log.warn(msg)
-        raise HTTPConflict()
-
-    if replace and attr_name not in attributes:
-        # Replace requires attribute exists
-        msg = f"Attempt to update missing attribute: {attr_name} "
-        msg += f"in obj_id: {obj_id}"
-        log.warn()
-        raise HTTPNotFound()
-
-    if replace:
-        orig_attr = attributes[attr_name]
-        create_time = orig_attr["created"]
-    else:
-        create_time = time.time()
 
-    # ok - all set, create attribute obj
-    attr_json = {"type": datatype,
-                 "shape": shape,
-                 "value": value,
-                 "created": create_time}
-    attributes[attr_name] = attr_json
+    # check for conflicts, also set timestamp
+    create_time = time.time()
+    new_attribute = False  # set this if we have any new attributes
+    for attr_name in items:
+        attribute = items[attr_name]
+        if attr_name in attributes:
```

Comment:
`attributes` should have a name that makes it clear that it's a list of attributes that already exist/are already in storage. Maybe `retrieved_attributes`? This would also be good for `attr_dict` in `POST_Attributes`

Reply:
The attributes are the current set of attributes, so I think I'll keep it the way it is.  Realize there's potential for confusion in other methods - will review.